lib/request-shim.js

```diff
@@ -55,7 +55,7 @@ function getOptions(options) {
         responseType: options.responseType || 'json'
     };
 
-    if (options.method !== 'GET') {
+    if (options.method !== 'GET' && options.body !== '') {
```

Comment:
Maybe I'm missing something, but in that case I'd say hat `body` and `json` overlaps in `got`.

For instance, if I set the following in `got`:

```
body: "foobar"
json: { "x": 1 }
```

what would be sent as payload of the outcoming request? Does `body` takes precedence over `json`? Or the other way around?

Reply:
As a side note, I understand that any existing usage of `json: true` of `json: false` in any call to the request-shim is wrong.

Maybe this should be checked in the existing code base of IOTAs, in order to early detect future problems.

Reply:
Regarding `json: true` usage across the agents and library, found coincidences at:

- `iotagent-node-lib`
- `iotagent-json`
- `lightweightm2m-iotagent`

<details>
  <summary>Screenshots</summary>

![Captura de pantalla 2022-01-17 a las 18 13 32](https://user-images.githubusercontent.com/40356341/149813731-660d9b8a-6e80-4613-aac8-18b39daa43e9.png)
![Captura de pantalla 2022-01-17 a las 18 14 05](https://user-images.githubusercontent.com/40356341/149813739-f1e1636e-8b7b-440b-b9bb-e29defb0bf9a.png)
![Captura de pantalla 2022-01-17 a las 18 15 40](https://user-images.githubusercontent.com/40356341/149813746-4290331a-804d-4e1c-95df-8289d6ae9b48.png)

</details>

Reply:
> Regarding `json: true` usage across the agents and library, found coincidences at:

Better to look for usages of `requests` function wrapper, to avoid losing cases due to slightly different pattersn (e.g. `json:true`, `json:\ntrue`, etc.).

Reply:
Looking at the screenshots:

- `deviceRequestWithoutPayload` is a GET - no body will be added anyway.
- `queryEntity` is a POST, but with `body` **and** `json` - it seems to be the case that `body` takes preference.
- most (if not all) of the remaining instances are in the unit tests

In all cases `json: true` should be deleted as it  is adding nothing.

Reply:
> but with body and json - it seems to be the case that body takes preference.

@jason-fox could you provie some reference in `got` documentation assuring this, please?

Reply:
> > but with body and json - it seems to be the case that body takes preference.
> 
> @jason-fox could you provie some reference in `got` documentation assuring this, please?

I finally found this at https://github.com/sindresorhus/got/blob/HEAD/documentation/2-options.md#body

> If `body` is specified, then the `json` or `form` option cannot be used.

What "cannot" means in this context? The process explodes with an exception raise? One of them is ignored (which would be the case empirically described by @jason-fox )?

`got` seems to be poorly documented... maybe we should ensure a defined precedence in our code in order not relay in an undocumented behaviour that may change in `got` new versions.

Reply:
SOF question trying to clarify this: https://stackoverflow.com/questions/70754880/undocumented-behaviour-related-with-body-json-and-form-parameters-in-got-node-m

Reply:
That could be added into the `request-shim.js` and documented in the code - something like:

```javascript
if (options.method === 'GET' || options.method === 'HEAD' || options.method === 'OPTIONS') {
        // Do nothing - Never add a body
} else if (options.body){
        // body takes precidence over json or form
        httpOptions.body = options.body;
}  else if (options.json){
        httpOptions.json = options.json;
} else if  (options.form){
        httpOptions.form = options.form;
}
```

Reply:
> That could be added into the `request-shim.js` and documented in the code - something like:
> 
> ```js
> if (options.method === 'GET' || options.method === 'HEAD' || options.method === 'OPTIONS') {
>         // Do nothing - Never add a body
> } else if (options.body){
>         // body takes precedence over json or form
>         httpOptions.body = options.body;
> }  else if (options.json){
>         httpOptions.json = options.json;
> } else if  (options.form){
>         httpOptions.form = options.form;
> }
> ```
> 
> You would then not be relying on the `got` library's precedence implementation.

Fixed in 7afd0ac